src/includes_normalize-win32.cc

```diff
@@ -28,16 +28,26 @@ namespace {
 
 bool InternalGetFullPathName(const StringPiece& file_name, char* buffer,
                              size_t buffer_length, string *err) {
-  DWORD result_size = GetFullPathNameA(file_name.AsString().c_str(),
-                                       buffer_length, buffer, NULL);
+  wchar_t w_buffer[_MAX_PATH];
+  wstring w_file_name = Utf8ToWide(file_name.AsString());
+  DWORD result_size =
+      GetFullPathName(w_file_name.c_str(), buffer_length, w_buffer, NULL);
```

Comment:
Can you elaborate on how

`void func(std::string const& str)
{
    std::wstring wstr = Utf8ToWide(str);
    WindowsFunctionHereW(wstr);
}
`

is better than

`void func(std::string const& str)
{
    WindowsFunctionHereA(str);
}
`
?

You're converting the `char` string to a `wchar_t` string manually, and then calling the wchar_t version of the windows function, but then converting the `wchar_t` that you get back from windows into a `char`.

Why is this better and/or different than simply calling the `char` version of the windows function?

I could see this change being an improvement over the `char` version if either (or both) of the two conditions were true, but they are not:

1) The string that you start with was `wchar_t`
2) The string that you get back was `wchar_t`

Because in either of those two situations, we save on the Windows API doing two string conversions (one for the input, one for the output). 

But in this situation, we're just lifting the string conversion from inside the windows API, into the Ninja code, which doesn't seem like a net-positive.

Reply:
Because char versions of Windows functions aren't UTF-8.

Reply:
Please elaborate?

Are you saying that because of the other changes to Ninja, this function also needs to change because the data stored in `std::string` is now globally, and exclusively, UTF8?

If that's the case, this PR needs to also add documentation about this change in semantics, and we'll need to do a thorough review of all the places where there might be direct string comparisons, and calls to functions like ::tolower, and others. Because the semantics of ascii and UTF8 are very different.

E.g. changing a UTF8 string to lower case requires a multi-byte aware function, because UTF8 characters that can be converted from upper / lower are not always single characters.

Reply:
Really, there's a good probability that we'll need to start linking against something like http://site.icu-project.org/home if Ninja is planning to support non-ascii strings anywhere at all.

Reply:
If a path contains a character not in the local charset, GetFullPathNameA won't work. At least that is my understanding of it, I'm no Windows expert.

Reply:
I don't understand what you mean.

UTF8 strings are universal. They don't have "char sets", but they do have "locales" that provide tweaks to how the UTF8 string is interpreted.

If this PR is changing the semantics of std::string objects in the Ninja codebase from ascii, to UTF8, then functions like `::tolower` must not be used. It will destroy the string data that it transforms, because UTF8 strings are multi-byte, and ::tolower is single-byte-at-a-time.

For example, this is the appropriate string function to call to "lowercase" a UTF8 string, in the ICU library:

https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/ustring_8h.html#ae3ed589f5ce0d550c228ee69bfed58fb

I strongly disagree with changing Ninja such that strings have UTF8 semantics, unless either of the two items below happen at the same time:

1) A full code review of the Ninja codebase is done to ensure all string manipulation is done in a UTF8 safe way.
2) A proper Unicode library is used, such as ICU.

Reply:
> UTF8 strings are universal. They don't have "char sets", but they do have "locales" that provide tweaks to how the UTF8 string is interpreted.

Yes, but the *A Windows functions have charsets. That's the problem.

> If this PR is changing the semantics of std::string objects in the Ninja codebase from ascii, to UTF8, then functions like ::tolower must not be used. It will destroy the string data that it transforms, because UTF8 strings are multi-byte, and ::tolower is single-byte-at-a-time.

`tolower` "works" (i.e. doesn't destroy) with UTF-8, since it doesn't touch non-ASCII bytes. All bytes in multi-byte characters are non-ASCII (since the first bit is set).

Reply:
To clarify, windows API suffixed with A, which are also the default for unsuffixed APIs when compiling without -D UNICODE, interpret strings as being in the "ANSI code page" (Windows terminology). The ANSI code page on any system depends on the language of the OS. For example, on an en-us Windows, the code page (aka encoding) of strings cannot represent japanese characters. This is why correct code can only be written using the -W functions, as file paths could contain any characters. The approach of this code review looks correct to me.

A library like ICU shouldn't be needed in code that does concatenations and most simple operations like that. It becomes needed when doing ordering/sorting, unicode normalization, unicode-aware comparisons, etc.

Note: I do not work on the Windows API and my statements here are my own based on research I've made and experience with similar issues.